common/list.c

```diff
@@ -1,134 +1,202 @@
 #include "list.h"
-#include <stdio.h>
+
 #include <stdlib.h>
+#include <stdbool.h>
 #include <string.h>
 
-list_t *create_list(void) {
-	list_t *list = malloc(sizeof(list_t));
-	if (!list) {
+list_t *list_new(size_t memb_size, size_t capacity) {
+	if (capacity == 0)
+		capacity = 8;
+
+	list_t *l = malloc(sizeof(*l));
+	if (!l) {
 		return NULL;
 	}
-	list->capacity = 10;
-	list->length = 0;
-	list->items = malloc(sizeof(void*) * list->capacity);
-	return list;
-}
 
-static void list_resize(list_t *list) {
-	if (list->length == list->capacity) {
-		list->capacity += 10;
-		list->items = realloc(list->items, sizeof(void*) * list->capacity);
+	l->capacity = capacity;
+	l->length = 0;
+	l->memb_size = memb_size;
+
+	l->data = malloc(memb_size * capacity);
+	if (!l->data) {
+		free(l);
+		return NULL;
 	}
+
+	return l;
 }
 
 void list_free(list_t *list) {
-	if (list == NULL) {
+	if (!list) {
 		return;
 	}
-	free(list->items);
+
+	free(list->data);
 	free(list);
 }
 
-void list_foreach(list_t *list, void (*callback)(void *item)) {
-	if (list == NULL || callback == NULL) {
-		return;
+static bool resize(list_t *list) {
+	if (list->length < list->capacity) {
+		return true;
 	}
-	for (int i = 0; i < list->length; i++) {
-		callback(list->items[i]);
+
+	size_t new_cap = list->capacity * 2;
+	void *data = realloc(list->data, list->memb_size * new_cap);
+	if (!data) {
+		return false;
 	}
-}
 
-void list_add(list_t *list, void *item) {
-	list_resize(list);
-	list->items[list->length++] = item;
+	list->data = data;
+	list->capacity = new_cap;
+	return true;
 }
 
-void list_insert(list_t *list, int index, void *item) {
-	list_resize(list);
-	memmove(&list->items[index + 1], &list->items[index], sizeof(void*) * (list->length - index));
-	list->length++;
-	list->items[index] = item;
-}
+void list_add(list_t *list, const void *data) {
+	if (!data || !list || !resize(list)) {
+		return;
+	}
+
+	size_t size = list->memb_size;
+	unsigned char (*array)[size] = list->data;
```

Comment:
>What with this syntax?

It's a pointer to a VLA. The syntax seems a bit tricky and clever, but it really cleans up the other parts of the code. We can treat it like a normal array, where pointer arithmetic will work properly.

>Also, why unsigned char over uint8_t?

I don't think uint8_t really adds much here. It's all basically "typeless" data that we're not going to touch. "unsigned char" seems good enough for a byte type.

>memcpy(&list->data[list->length], ...) should work the same, right?

No, because list->data is a void *, so we can't do pointer arithmetic. It would require casting to a char type and then doing the offsets yourself.